File: Backtracking/NQueen.cpp
#include <bits/stdc++.h>
using namespace std;
#define fr(a,b) for(i=a;i<b;i++)
#define rep(a,b) for(i=a;i>=b;i--)
#define input(a,n) fr(i,n) cin>>a[i]
#define print(a,n) fr(i,n) {cout << a[i] << " ";} cout << endl
#define lli long long int
#define ull unsigned long long

int m[11][11];

bool place(int row,int col,int n)
{
    //check above

    for(int i = row;i>=0;i--)
    {
        if(m[i][col]==1)
            return false;
    }

    //check left diagonal

    for(int i =row, j = col;i>=0,j>=0;i--,j--)
    {
        if(m[i][j]==1)
            return false;
    }

    //check right diagonal

    for(int i= row, j = col;i>=0,j<n;i--,j++)
    {
        if(m[i][j]==1)
        {
            return false;
        }
    }
    //this will return whether we can place
    //queen at that block or not
    return true;
}



void nQueen(int n,int row)
{
    if(row>=n)
    {
        cout<<"==========================="<<endl;
        for(int i = 0;i<n;i++)
        {
            for(int j = 0;j<n;j++)
            {
                cout<<m[i][j]<<" ";
            }
            cout<<endl;
        }
        cout<<"==========================="<<endl;
    }
    for(int i = 0;i<n;i++)
    {
        if(place(row,i,n))
        {
            m[row][i] = 1;
            nQueen(n,row+1);
            m[row][i]=0;
        }
    }
}



int main()
{
    int n;
    cin>>n;
    nQueen(n,0);
    return 0;
}

File: Backtracking/nqueen.py
def place(row, col, n, m):
    
    # Check above row of the same column
    for i in range(row, -1, -1):
        if m[i][col] == 1:
            return False

    # Check left upper diagonal
    for i, j in zip(range(row, -1, -1),
                    range(col, -1, -1)):
        if m[i][j]:
            return False
    
    # Check right upper diagonal
    for i, j in zip(range(row, -1, -1),
                    range(col, N)):
        if m[i][j]:
            return False

    return True


def nQueen(n, row, m):
    # print(row)
    if row >= n:
        # print("Entering print condition")
        for i in range(0, n):
            for j in range(0, n):
                print(m[i][j], end=" ")
            print()

        return True
    for i in range(0, n):
        if place(row, i, n, m):
            m[row][i] = 1
            if nQueen(n, row + 1, m) == True:
                return True
            m[row][i] = 0
    return False

board = [[0 for y in range(0, 11)] for x in range(0, 11)]
N = int(input())
if nQueen(N, 0, board) == False:
    print("No Solution Exist")
else:
    print("Solution")
